### PR TITLE
New logic for default parameters in IA classes

### LINKED ISSUE #3069: Rethink set_default_params()
The bonded IA classes in Python have a method `set_default_params()` that is _supposed_ to set default values. It is unclear to me what this method is actually doing in the current state, because:
* it's followed by a call to `_params.update()` that will override them by those stored in the core
  https://github.com/espressomd/espresso/blob/c2e8604c4d914b70616133be32989f9203baedaf/src/python/espressomd/interactions.pyx#L1904-L1907
* it's often setting a default value to required parameters, which by definition can't have a default
  https://github.com/espressomd/espresso/blob/c2e8604c4d914b70616133be32989f9203baedaf/src/python/espressomd/interactions.pyx#L2125-L2135
* they don't always provide sensible values, like this angle potential where an angle of 0 or pi would lead to NaN values in the force functions because of the division by `sin(phi) = 0`:
  https://github.com/espressomd/espresso/blob/c2e8604c4d914b70616133be32989f9203baedaf/src/python/espressomd/interactions.pyx#L2975-L2979
* one of them has a typo
  https://github.com/espressomd/espresso/blob/c2e8604c4d914b70616133be32989f9203baedaf/src/python/espressomd/interactions.pyx#L2549-L2553
  which is really confusing:
  ```python
  di = espressomd.interactions.Dihedral(mult=2, bend=5, phase=0)
  # mult should be 2, but it shows 1 on the left
  di.params
  ```
  ```
  {"mult'": 1, 'bend': 5, 'phase': 0, 'mult': 2}
  ```

---

This method should return a dict of sensible default values, only be called once in the class constructor, and its keys should be checked for typos. To do all at once, one could replace this line:
https://github.com/espressomd/espresso/blob/c2e8604c4d914b70616133be32989f9203baedaf/src/python/espressomd/interactions.pyx#L1862
with e.g.:
```python
params = get_default_params()
params.update(kwargs)
self.params = params
```

### PR DESCRIPTION
**Pull request by jngrad**
Fixes #3069

Change internals of bonded/non-bonded IA Python classes:
- `valid_keys()` and `required_keys()` now return sets instead of tuples/lists/strings (strings can lead to cryptic error messages since they are iterable)
- `default_params()` and `set_default_params()` now return values only for optional parameters
- wrong keys (typos or unknown keys) in optional parameters were fixed
- added tests for a few common IAs to check the new syntax and verify default parameters have the correct name and don't overlap with required parameters, which can't have default values

### CODE REVIEW

File: src/python/espressomd/interactions.pyx [hunk 1]
```diff
@@ -487,18 +479,7 @@ IF LENNARD_JONES_GENERIC == 1:
...
-                "cutoff": 0.,
-                "shift": 0.,
-                "offset": 0.,
-                "e1": 0,
-                "e2": 0,
-                "b1": 0.,
-                "b2": 0.,
-                "delta": 0.,
-                "lam": 0.}
+            return {"delta": 0., "lam": 0.}
```
Comment (KaiSzuttor) [inline src/python/espressomd/interactions.pyx]: why are only those two parameters default ones? Shouldn't we reflect the facts from the core here?

Comment (jngrad) [reply src/python/espressomd/interactions.pyx]: * `lam` is now 1 by default to reflect the core
* this line is now guarded by `LJGEN_SOFTCORE`
* keeping them as optional allows this generic LJ class to behave the same way when `LJGEN_SOFTCORE` is enabled, unless the user provides custom values for `lam` and `delta`

**Comment by jngrad** [comment]
@KaiSzuttor is anything missing?

**Review by KaiSzuttor — APPROVED** [review]
LGTM in the future we probably want to have constructors for the interactions in the core. Or at least on the python level.

**Comment by jngrad** [comment]
Agreed. The Python class should only take care of checking variable types, the core should take care of checking values. Core constructors are planned in #3037.

**Comment by jngrad** [comment]
bors r=KaiSzuttor

**Comment by jngrad** [comment]
bors cancel

**Comment by jngrad** [comment]
bors r=KaiSzuttor

**Pull request merged**